Constructing a custom element should immediately execute microtasks if there are no scripts

rniwa opened the pull request:
#### e7be2359033b5a89111f70b799953e17421eeede
<pre>
Constructing a custom element should immediately execute microtasks if there are no scripts
<a href="https://bugs.webkit.org/show_bug.cgi?id=188629">https://bugs.webkit.org/show_bug.cgi?id=188629</a>

Reviewed by Chris Dumez.

Perform a microtask check point when constructing a custom element with no script execution in the parser.

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/microtasks-and-constructors-expected.txt:
* Source/WebCore/bindings/js/JSCustomElementInterface.cpp:
(WebCore::JSCustomElementInterface::constructElementWithFallback):
(WebCore::JSCustomElementInterface::tryToConstructCustomElement):
(WebCore::constructCustomElementSynchronously):
* Source/WebCore/bindings/js/JSCustomElementInterface.h:
* Source/WebCore/html/parser/HTMLDocumentParser.cpp:
(WebCore::HTMLDocumentParser::runScriptsForPausedTreeBuilder):

Canonical link: <a href="https://commits.webkit.org/252190@main">https://commits.webkit.org/252190@main</a>
</pre>

cdumez commented on Source/WebCore/bindings/js/JSCustomElementInterface.h:
```diff
@@ -58,7 +58,7 @@ class JSCustomElementInterface : public RefCounted<JSCustomElementInterface>, pu
     }
 
-    Ref<Element> constructElementWithFallback(Document&, const AtomString&);
+    Ref<Element> constructElementWithFallback(Document&, const AtomString&, bool parserConstructElementWithEmptyStack = false);
```
We should use a enum class instead of a boolean parameter.

rniwa replied:
so `enum class ParserConstructElementWithEmptyStack { Yes, No }`?

cdumez replied:
Ideally:
`enum class ParserConstructElementWithEmptyStack: bool { No, Yes };`

I'm not a huge fan of the name but I don't have a better proposal at the moment.

cdumez reviewed (approved):
LGTM

rniwa committed:
Constructing a custom element should immediately execute microtasks if there are no scripts
https://bugs.webkit.org/show_bug.cgi?id=188629

Reviewed by Chris Dumez.

Perform a microtask check point when constructing a custom element with no script execution in the parser.

* LayoutTests/imported/w3c/web-platform-tests/custom-elements/microtasks-and-constructors-expected.txt:
* Source/WebCore/bindings/js/JSCustomElementInterface.cpp:
(WebCore::JSCustomElementInterface::constructElementWithFallback):
(WebCore::JSCustomElementInterface::tryToConstructCustomElement):
(WebCore::constructCustomElementSynchronously):
* Source/WebCore/bindings/js/JSCustomElementInterface.h:
* Source/WebCore/html/parser/HTMLDocumentParser.cpp:
(WebCore::HTMLDocumentParser::runScriptsForPausedTreeBuilder):

Canonical link: https://commits.webkit.org/252190@main
